### ISSUE TITLE
Scheduling Problem with LocalExecutor

### ISSUE THREAD

**Issue by r39132** [apache/airflow#341]
I'm running the LocalExecutor. 

On recent versions of the code (I'm seeing this on and after 1.4.1 all the way up till the latest code in master), I'm running into some thing strange with `backfill` and LocalExecutor. 

Essentially, no work is being picked up or done!

Here's what I see repeatedly on multiple (2) servers, one running latest master, the other running 1.4.1. Earlier versions of the code did not have this problem. 

Kicking off a backfill, I see the following on stdout

```
INFO:root:Running airflow backfill ep_demo -t aggregate_data_spark_job -s 2015-08-20 -e 2015-08-20
2015-08-31 22:30:00,577 INFO - Filling up the DagBag from /Users/siddharth/Projects/airflow/dags
2015-08-31 22:30:00,583 INFO - Importing /Users/siddharth/Projects/airflow/dags/ep_demo_1.py
2015-08-31 22:30:00,636 INFO - Loaded DAG <DAG: ep_demo>
2015-08-31 22:30:00,699 INFO - Adding to queue: airflow run ep_demo aggregate_data_spark_job 2015-08-20T00:00:00   --local -i    -sd DAGS_FOLDER/ep_demo_1.py -s 2015-08-20T00:00:00 
2015-08-31 22:30:05,643 INFO - [backfill progress] waiting: 1 | succeeded: 0 | kicked_off: 1 | failed: 0 | skipped: 0 
2015-08-31 22:30:05,644 INFO - LocalWorker running airflow run ep_demo aggregate_data_spark_job 2015-08-20T00:00:00   --local -i    -sd DAGS_FOLDER/ep_demo_1.py -s 2015-08-20T00:00:00 
2015-08-31 22:30:10,641 INFO - [backfill progress] waiting: 1 | succeeded: 0 | kicked_off: 1 | failed: 0 | skipped: 0 
Logging into: /Users/siddharth/Projects/airflow/logs/ep_demo/aggregate_data_spark_job/2015-08-20T00:00:00
2015-08-31 22:30:15,641 INFO - [backfill progress] waiting: 1 | succeeded: 0 | kicked_off: 1 | failed: 0 | skipped: 0 
2015-08-31 22:30:20,640 INFO - [backfill progress] waiting: 1 | succeeded: 0 | kicked_off: 1 | failed: 0 | skipped: 0 
2015-08-31 22:30:25,645 INFO - [backfill progress] waiting: 1 | succeeded: 0 | kicked_off: 1 | failed: 0 | skipped: 0 
2015-08-31 22:30:30,642 INFO - [backfill progress] waiting: 1 | succeeded: 0 | kicked_off: 1 | failed: 0 | skipped: 0 
2015-08-31 22:30:35,645 INFO - [backfill progress] waiting: 1 | succeeded: 0 | kicked_off: 1 | failed: 0 | skipped: 0 
2015-08-31 22:30:40,649 INFO - [backfill progress] waiting: 1 | succeeded: 0 | kicked_off: 1 | failed: 0 | skipped: 0 
2015-08-31 22:30:45,648 INFO - [backfill progress] waiting: 1 | succeeded: 0 | kicked_off: 1 | failed: 0 | skipped: 0  
```

The Tree View shows nothing happening for the 8/20/2015 run at all.

<img width="925" alt="screenshot 2015-08-31 22 38 21" src="https://cloud.githubusercontent.com/assets/581734/9596964/09d3f27c-5031-11e5-9653-c113ec0b4315.png">

In the logs, I see 

```
[2015-08-31 22:30:11,400] {models.py:104} INFO - Filling up the DagBag from /Users/siddharth/Projects/airflow/dags/ep_demo_1.py
[2015-08-31 22:30:11,402] {models.py:171} INFO - Importing /Users/siddharth/Projects/airflow/dags/ep_demo_1.py
[2015-08-31 22:30:11,454] {models.py:223} INFO - Loaded DAG <DAG: ep_demo>
[2015-08-31 22:30:17,164] {models.py:104} INFO - Filling up the DagBag from /Users/siddharth/Projects/airflow/dags/ep_demo_1.py
[2015-08-31 22:30:17,166] {models.py:171} INFO - Importing /Users/siddharth/Projects/airflow/dags/ep_demo_1.py
[2015-08-31 22:30:17,219] {models.py:223} INFO - Loaded DAG <DAG: ep_demo>
[2015-08-31 22:30:17,223] {models.py:834} INFO - 
--------------------------------------------------------------------------------
New run starting @2015-08-31T22:30:17.223143
--------------------------------------------------------------------------------
[2015-08-31 22:30:17,227] {models.py:845} INFO - Queuing into pool ep_data_pipeline_spark_tasks_only
```

I then look at the Pool view and in fact see nothing queued in the pool at all

<img width="892" alt="screenshot 2015-08-31 22 39 58" src="https://cloud.githubusercontent.com/assets/581734/9596991/7723df86-5031-11e5-9574-3b65b8e611d6.png">

All the while, the schedulers are all reporting the following in their own terminal window

```
2015-08-31 22:42:08,346 - root - INFO - Getting latest instance for all task in dag ep_demo
2015-08-31 22:42:08,379 - root - INFO - Done queuing tasks, calling the executor's heartbeat
2015-08-31 22:42:08,379 - root - INFO - Loop took: 0.042677 seconds
2015-08-31 22:42:13,349 - root - INFO - Done queuing tasks, calling the executor's heartbeat
2015-08-31 22:42:13,349 - root - INFO - Loop took: 0.01327 seconds
2015-08-31 22:42:18,346 - root - INFO - Getting latest instance for all task in dag ep_demo
2015-08-31 22:42:18,379 - root - INFO - Done queuing tasks, calling the executor's heartbeat
2015-08-31 22:42:18,379 - root - INFO - Loop took: 0.043476 seconds
2015-08-31 22:42:23,349 - root - INFO - Done queuing tasks, calling the executor's heartbeat
2015-08-31 22:42:23,349 - root - INFO - Loop took: 0.014791 seconds
```

Why is my DAG not being run? I can run it using `airflow test`, but anything using the LocalExecutor does not work. 

My code, or at least part of it, is shown below : 

```
from airflow import DAG, utils
from airflow.operators import BashOperator, BranchPythonOperator, DummyOperator, EmailOperator, PythonOperator
from datetime import date, datetime, time, timedelta
from ep_data_pipeline_utils import check_for_sqs_message_branch_condition, ENV, \
    import_aggregate_lookback_days, import_airflow_start_date_as_lookback_days, \
    import_ep_pipeline_success_email_dl, import_spark_cluster_ip, \
    send_email_notification_flow_successful, send_sns_notification_no_spark_data, \
    wait_for_new_data_in_db, wait_for_empty_queue


START_DATE = datetime.combine(datetime.today() -
                              timedelta(int(import_airflow_start_date_as_lookback_days)),
                              datetime.min.time())
default_args = {
    'owner': 'sanand',
    'depends_on_past': True,
    'pool': 'ep_data_pipeline',
    'start_date': START_DATE,
    'email': [import_ep_pipeline_success_email_dl],
    'email_on_failure': True,
    'email_on_retry': True,
    'retries': 0,
    'retry_delay': timedelta(seconds=60)
}
dag = DAG('ep_demo' , default_args=default_args)


# Operator DAG Design
#1. Generate Parquet files -- this also consumes new upload files and moves them into archived
#2. Run Aggregate Spark Job -- this generates aggregate data from Parquet files
#3. Check SQS for new messages -- if no messages, branch --> send email
#4. Check the DB for new rows within exec_date - 1d <--> exec_date
#5. Wait for SQS to drain
#6. Send an email that the flow completed successfully
```

**Comment by mistercrunch** [comment]
Can you double check that `1.4.0` or `1.3.0` did not have this issue by running `pip install airflow==1.4.0` and running your test?

I run the unit tests very often and they cover backfill on a LocalExecutor, so there has to be something else going on here.

Another thing to try to narrow this down would be to momentarily switch depends_on_past to False and see if it helps. The logic is supposed to not apply depends_on_past on the start_date of a task or a backfill, but there may be an issue around that...

**Comment by kmmats** [comment]
I'm having a similar issue but on v1.3.0 -- the output logs are hanging (mid print statement) and the kicked off job never completes -- although based on the anticipated outputs of the script, the script itself does complete.

also using Local Executor

**Comment by r39132** [comment]
OK. I found the cause.

We run airflow (webserver and scheduler) on one of our EC2 machines called ETL-00. We auto-scale this instance for one of the processes (importers) running on this machine. This machine runs many processes that don't need auto-scaling, such as the airflow webserver and scheduler, so we need to ensure those processes are stopped on the "replicated" auto-scaled instances. It turns out that we were not killing or stopping the airflow scheduler fast enough on those replicas, so the replicas were picking up tasks not meant for them. And, since the user-data would kill the process on the auto-scaled machines, the work was not completing. 

I'm closing this ticket.

**Comment by mistercrunch** [comment]
Good catch. I admit it can get tricky to read through what is happening in a distributed environment. I noticed people get tangled, post about it, and it's really hard to help out from outside the environment.

One thing that could help is better event logging from all involved machines / processes. If we logged more events like "scheduler started", "scheduler ended", "web server started", "read msg from queue", along with the hostname, timestamp, environment variables, ... that could really help people people sys admins on a small cluster of Airflow machines.
